### PR TITLE
Adding support for pgoutput in Producer

### PR DESCRIPTION
**Pull request by Shwetabhk**
**Title**: Update PostgreSQL Dockerfile and Event Producer Configuration

**Description**:

This Pull Request (PR) includes several updates and improvements to the codebase. The changes are summarized below:

**Changes Made**:

- In the `common/qconnector/rabbitmq_connector.py` file, the `publish` method has been updated to optionally compress the message payload before publishing it to RabbitMQ. This change introduces a new parameter, `compress_message`, which, when set to `True`, compresses the payload. This can help optimize the use of bandwidth when publishing messages.

- A new utility class, `common/utils.py`, has been introduced. This class includes two static methods:
  - `convert_bytes_to_int`: Converts a byte sequence to an integer, following the big-endian byte order.
  - `convert_bytes_to_utf8`: Converts a byte sequence to a UTF-8 encoded string.

- In the `dev/postgres/fixtures/fixture-1.sql` file, a new publication named `events` has been created for all tables. This publication can be used for logical replication.

- The `docker-compose.yaml` file has been updated to include a new environment variable, `PGOUTPUTPLUGIN`, with the value `wal2json`. This environment variable specifies the PostgreSQL output plugin to be used.

- In the `producer/event_producer.py` file, several changes have been made:
  - Connection pooling using `psycopg2.pool.SimpleConnectionPool` has been introduced for PostgreSQL connections, providing better resource management and performance.
  - The logic for creating a replication slot has been updated to be more flexible and compatible with different output plugins.
  - Two new message processors, `wal2json_msg_processor` and `pgoutput_msg_processor`, have been added to handle different types of messages based on the chosen output plugin.
  - The `__get_table_name` method has been introduced to retrieve the table name from a relation ID.
  - Messages are published without compression when using the `pgoutput` output plugin.

**Testing**:

- The changes have been tested locally to ensure that they work as expected.
- Specific testing scenarios include verifying the Dockerfile changes, initialization script behavior, and message processing for different PostgreSQL output plugins.

### CODE REVIEW

File: producer/event_producer.py [hunk 1]
```diff
@@ -118,12 +136,63 @@ def msg_processor(self, msg):
         msg.cursor.send_feedback(flush_lsn=msg.data_start)
         self.check_shutdown()
 
+    def __get_table_name(self, relation_id: int) -> str:
```
Comment (snarayanank2) [inline producer/event_producer.py:139]: add caching to avoid repeated calls

File: producer/event_producer.py [hunk 2]
```diff
@@ -118,12 +136,63 @@ def msg_processor(self, msg):
...
+        message_type = msg.payload[:1].decode('utf-8')
+
+        # Convert bytes to int
+        relation_id = parser_utils.convert_bytes_to_int(msg.payload[1:5])
+
+        if message_type in ['I', 'U', 'D']:
+            operation_type = 'INSERT' if message_type == 'I' else 'UPDATE' if message_type == 'U' else 'DELETE'
+            table_name = self.__get_table_name(relation_id)
+
+            logger.info(f'{operation_type} Change occurred on table: {table_name}')
```
Comment (snarayanank2) [inline producer/event_producer.py:171]: change to debug

**Review by snarayanank2 — CHANGES_REQUESTED** [review]

File: common/qconnector/rabbitmq_connector.py [hunk 3]
```diff
@@ -8,14 +8,15 @@
 
 
 class RabbitMQConnector(QConnector):
-    def __init__(self, rabbitmq_url, rabbitmq_exchange, queue_name=None, binding_keys=None):
+    def __init__(self, rabbitmq_url, rabbitmq_exchange, queue_name=None, binding_keys=None, use_compression=True):
```
Comment (Shwetabhk) [inline common/qconnector/rabbitmq_connector.py:11]: Named variable as `use_compression` to later use the same variable not no decompress pgoutput messages

File: dev/postgres/fixtures/fixture-1.sql [hunk 4]
```diff
@@ -4,3 +4,5 @@ create table users (
 );
 
 alter table users replica identity full;
+
+create publication events for all tables;
```
Comment (aditya999123) [inline dev/postgres/fixtures/fixture-1.sql:8]: this should be configurable

File: producer/event_producer.py [hunk 5]
```diff
@@ -35,37 +42,52 @@ def __init__(self, *, qconnector_cls, event_cls, pg_host, pg_port, pg_database,
...
+                'format-version': 2,
+                'include-types': True,
+                'include-lsn': True
+            }
+            if self.__pg_tables and len(self.__pg_tables) > 0:
+                options['add-tables'] = self.__pg_tables
+        else:
+            options = {
+                'proto_version': 1,
+                'publication_names': 'events'
```
Comment (aditya999123) [inline producer/event_producer.py:78]: configurable

File: producer/event_producer.py [hunk 6]
```diff
@@ -35,37 +42,52 @@ def __init__(self, *, qconnector_cls, event_cls, pg_host, pg_port, pg_database,
...
             user=self.__pg_user,
             password=self.__pg_password,
             connection_factory=self.__pg_connection_factory
         )
-        self.__db_cur = self.__db_conn.cursor()
+        self.__db_cur = self.__db_conn_pool.getconn().cursor()
+
+        self.__replication_cursor = self.__db_conn_pool.getconn().cursor()
+
+        decode = True
```
Comment (aditya999123) [inline producer/event_producer.py:65]: move this inside wal2json condition

File: producer/event_producer.py [hunk 7]
```diff
@@ -118,12 +140,63 @@ def msg_processor(self, msg):
         msg.cursor.send_feedback(flush_lsn=msg.data_start)
         self.check_shutdown()
 
+    @cachetools.cached(cache={}, key=lambda self, relation_id: cachetools.keys.hashkey(relation_id))
```
Comment (aditya999123) [inline producer/event_producer.py:143]: don't follow, what is expiry here ?
what happens when tablename changes - unlikely but still possible

**Review by aditya999123 — APPROVED** [review]

File: dev/postgres/fixtures/fixture-1.sql (hunk 4 above)
Comment (Shwetabhk) [reply dev/postgres/fixtures/fixture-1.sql:8]: This is just a fixute, I've added support for passing publication name in the environment

File: producer/event_producer.py [hunk 8]
```diff
@@ -118,12 +147,98 @@ def msg_processor(self, msg):
...
+            operation_type = 'INSERT' if message_type == 'I' else 'UPDATE' if message_type == 'U' else 'DELETE'
+            table_name = self.__get_table_name(relation_id)
+
+            if table_name in self.__pg_tables:
+                logger.debug(f'{operation_type} Change occurred on table: {table_name}')
+                logger.debug(f'{operation_type} Change occurred at LSN: {msg.data_start}')
+
+                payload = {
+                    'schema': self.__get_schema(relation_id),
+                    'payload': base64.b64encode(msg.payload).decode('utf-8')
```
Comment (Shwetabhk) [inline producer/event_producer.py:217]: encoding to base64 and decoding to convert to string
Bytes data was not being compressed

File: producer/event_producer.py [hunk 9]
```diff
@@ -1,29 +1,43 @@
...
 from common.log import get_logger
 
+from common.utils import DeserializerUtils as parser_utils
+
+
 logger = get_logger(__name__)
 
+# 12 hour cache
+# Maximum of 1024 entries
+table_name_cache = cachetools.TTLCache(maxsize=1024, ttl=43200)
```
Comment (aditya999123) [inline producer/event_producer.py:25]: can move ttl to env variable, easier to change later on
the 12 hour of tablename looks too high

File: producer/event_producer.py [hunk 10]
```diff
@@ -118,12 +147,98 @@ def msg_processor(self, msg):
...
+                (relation_id,)
+            )
+            schema_name, table_name = self.__db_cur.fetchone()
+            return f'{schema_name}.{table_name}'
+        except Exception:
+            logger.exception("Failed to get table name.")
+            raise
+    
+    @cachetools.cached(cache=schema_cache, key=lambda self, relation_id: cachetools.keys.hashkey(relation_id))
+    def __get_schema(self, relation_id) -> dict:
```
Comment (aditya999123) [inline producer/event_producer.py:174]: get_table_schema

**Review by aditya999123 — APPROVED** [review]

**Pull request merged**